doc/wfc.1
.TH "wfc" "1" "R1906" "Thomas Maier-Komor" "console utility"
.SH "NAME"
WFC \- wire format compiler
.SH "SYNTAX"
.LP 
wfc [\fIoptions\fP] <file>
.SH "DESCRIPTION"
.LP 
WFC translates protocol descriptions to C++ source files. The generated
header files provide class descriptions for serializing and
deserializing data accoring to the protocol description.

WFC syntax has been derived from protocol buffers, but provides
optimizations for embedded systems.
.SH "OPTIONS"
.LP 
\fB\-i\fR <\fIfilename\fP>
Use \fIfilename\fP as input instead of the standard input (needs to be
given for multi volume support). If \fIfilename\fP is \-, input is read
from standard input.
.br
\fB\-m\fR <\fImessage\fP>
Generate code starting with message \fImessage\fP. Depending messages
will be generated if necessary.
.br
\fB\-t\fR <\fItarget\fP>
Use option set \fItarget\fP as specified in the source file to generate
the code.
.br
\fB\-o\fR <\fIout_name\fP>
Generate files named \fIout_name\fP.cpp and \fout_name\fP.h.
.br
\fB\-l\fR
Genrate wfc library only. This is useful if you want to share the
library among multiple source files.
.br
\fB\-v\fR
Operate in verbose mode.
.br
\fB\-s\fR
Generate subclasses instead of mangled class names.
.br
.SH "OPTIONS AND DEFAULT VALUES"
Following options with associated default values are available. Option names
are case insensitive. for options can be overridden on the command line or in
the .wfc source file.
.br
\fIOptimize\fP: valid values are review, size, speed; default is review
.br
\fIwfclib\fP: valid values are static, inline, extern. Static means generate
shareable functions as static/module local functions. Inline generates the
functions in the header file. Extern forces the use of externally defined
functions. For extern usage, the library can be generated with option \fI-l\fP
(see above).
\fIUnkownField\fP: valid values are assert, skip; default is skip. This option
defines the handling of unkonwn message field tags during parsing. Usually you
want to skip unknown fields.
.br
\fIwfclib\fP: valid values are static, inline, extern; default is static
.br
\fIendian\fP: valid values are unknown, little, big; default is unknown.
Use little endian optimization only for targets that are able to perform
an unaligned access. Big endian and unknown endian are currently implemented
without any difference.
.br
\fIwithEqual\fP: valid values are true, false; default is true. If this
option is enabled the operator == will be generated.
.br
\fIwithUnequal\fP: valid values are true, false; default is true. If this
option is enabled the operator != will be generated.
.br
\fIMutableType\fP: return type of mutable_* member functions. This can be
either a pointer or a reference. Accordingly, the agrument must be either
"pointer" or "reference". Default is "pointer".
.br
\fIErrorHandling\fP: what to do in case of errors that are detected during
reading data. Valid methods are: cancel (i.e. return with an error code),
assert (i.e. call abort), throw (throw an exception).
.br
\fIVarIntBits\fP: Number of bits to use for variable length integer
types. Default is 64, set to 32 or 16 on machines with smaller native
types, if the value range of all fields allows it.
.br
\fIIntSize\fP: default number of bits for data-types unsigned, signed, and int.
.br
\fIheader\fP: include a header file in the generated code. This option
can be used multiple times.
.br
\fIheader\fP: This option can be used multple times and tells wfc to add an
include directive for the specific header file into the generated code. This is
needed if you intend to use custem string types.
.br
\fInamespace\fP: Set the namespace for which the code shall be generated. This
might be necessary, if you want to integrate multiple WFC definitions with
conflicting settings. See wfclib example.
.br
\fItoASCII\fP: name of toASCII method. This is disabled per default, as linking
with wfc_support.cpp is otherwise necessary.
.br
\fIBaseClass\fP: base class for generated message classes.
.br
\fIstreamtype\fP: Set class name to use for output streams. Default is
std::ostream.
.br


.SH "MESSAGE OPTIONS"
\fISortMembers\fP: member variables are sorted according to the given sorting
scheme. Valid arguments are: id (default), name, type, size, unsorted.
.br
\fIinline\fP: Comma separated lists of methods to inline. Defaults
depend on Optimization level. Valid arguments are get, set, has.
.br
\fImembers\fP (BETA): message members can be either 'regular' or 'virtual'.
When members is set to virtual the member variables are accessed via 'virtual
functions and no member variables will be included in the class definition.


.SH "FIELD OPTIONS"
Options that are applicable to elements of messages.
.br
\fIused\fP: valid values are true, false; default is true. If this
option is set to false, no code will be generated for handling this
field.
.br
\fIstringtype\fP: Data type to use for strings. Default is std::string.
Set to "C" to use C string (const char *).
.br
\fIarraysize\fP: Integer value of number of elements to allocate for a
repeated type. If set to 0 (default), a std::vector will be used
instead. Use if maximum number of elements is known, if you want to
avoid impact of std::vector.
.br
\fIunset\fP: sets an in range value that will be interpreted as the value being
not set. I.e. has_xx will return false if its value is the \fIunset\fP value.
This eleminates the requirement for a dedicated bit to implement the related
has_ methods. The clear methods also use the unset value to reset the
field.
.br
\fIdefault\fP: default value that will be assigned during object
initialization and when clearing the field value. If the field is an
optional field, the field is presented as unset nevertheless. Note also
that if unset is set to a different value, the value after object
construction and after calling clear() will differ.
.br
\fIto\_ascii\fP: specifies a function that will be used for generating an
ASCII representation for its value. Like this you can produce readable outputs
for binary coded words. E.g. the decimal dot representiation of an IPv4
address. See the examples directory, to see how it works.
.br
\fIto\_json\fP: specifies a function that will be used for generating a
JSON representation for its value. It works like \fIto\_ascii\fP, but is
complementary available, because JSON output might require certain characters
to be escaped differently when compared to ASCII output.
.br
\fIusage\fP: usage can be regular, deprecated, and obsolete. Default is
regular. If usage is set to deprecated, data will not be serialized, but
existing serialzed data will be read into memory. This is for fading out use of
a certain field. Once you set the usage to obsolte, the data will neither be
written or read anymore. Then the field specification is just there for
blocking against future use, and documenting the fact that it was once there.
.br
\fISetByName\fP: name of function that can be used to assign members by name with
ASCII input. This does not work for repeated types and bytes types. Enum types
are limited to parsing their integer value for now. Only individual message
fields can be set; i.e. parsing of complete messages like the ASCII output is
out of scope. The function performs no input validity checking.
Subelements of message fields can be selected with dot notation as in C:
e.g. msg_a.field_b.element
.br
\fIstorage\fP (BETA): like option \fImembers\fP this option sets the instantiation
style of the data for the relevant field. I.e. if set to true, all member
functions will be pure virtual, and no data will be instantiated. This
option is limited to the relevant field is set, while in contrast the option
\fImembers\fP at the message level will set the default for all the members of
the message.


.SH "EXAMPLES"
.LP 
To generate header and c++ source files accoring to a wfc protocol
description, just run:
.LP 
wfc <sourcefile>.wfc
.LP 
To generate the code fot a specific target option set, use:
wfc -t target source.wfc
.LP
See the examples subdirectory for complete examples of protocol
descriptions, and option usage.
.LP 
.SH "AUTHORS"
.LP 
Thomas Maier\-Komor <thomas@maier\-komor.de>
.SH "DONATIONS"
.LP
If you like this software, and use it for production purposes in your
company, please consider making a donation to support this work. 
You can donate directly via PayPal to the author's e-mail address
(thomas@maier\-komor.de).
.SH "HOMEPAGE"
.LP
http://www.maier\-komor.de/wfc.html
.SH "LICENSE"
.LP
This software is published under GNU General Public License V3. See file
LICENSE for details.
.SH "SEE ALSO"
.LP 
protoc(1)


ChangeLog
Release R2003:
==============
- enhancement: added option to set field ascii parsing function
- enhancement: ascii output should include unset fields
- enhancement: clear() should reset the member to the default value
- enhancement: support to clear field with setByName(value = 0)
- enhancement: support to add element with setByName("element[+]",...)
- fix: assertion failure with multiple field options
- fix: field options clear and unset must play together
- fix: static storage initialization broken

Release R2002:
==============
- enhancement: getMaxSize returns SIZE_MAX if size is unlimited
- enhancement: output custom declarations to .cpp
- enhancement: support for custom JSON value generation
- enhancement: storage type specification
- enhancement: repeated types should have a const& get
- enhancement: added concept for reserving ids
- enhancement: set value by name from ASCII (restrictions see man page)
- enhancement: removed dependency to wfc_support.* for ASCII output
- enhancement: added support for custom template functions
- documentation updates (ongoing)
- fix: unset value should be used as default, if default is not set
- fix JSON output for 8 bit integers
- fix related to subclass naming
- fixes and updates for alternate stringtypes

Release R2001:
==============
- enhancement: optimized JSON generation code
- enhancement: added fixed size string header and usage example
- updated incomplete astring.h
- updated example syscfg
- documentation updates (still incomplete)
- added new examples and updated existing ones
- minor bugfixes

Release R1907:
==============
- enhancement: messages and enums also need keyword safeguards
- enhancement: new member sortings: name, type, unsorted
- fix: omit operator declaration for ==, !=, if suppressed
- POSIX compliant options parsing to support Solaris and *BSD
- added man-page

Release R1905:
==============
- fix: generated code with fixed8/fixed16 may not compile
- fix: invalid handing of int8 in some cases
- fix: remove superfluous duplicate parenthesis
- enhancement: use emplace_back instead of resize
- enhancement: safeguard identifiers against conflicts with keywords
- change: disable Sink support per default to avoid include dependency

Release R1902:
==============
- first public release
- use option -h to query available options


Versions below are unpublished prior releases that were used for hostscope and
other private projects.

Version 0.6.0:
==============
- added null/ff termination support

Version 0.5.0:
==============
- added JSON generation
- added XML generation
- added support for invalid value based has_ methods
- added option to return referene instead pointer from mutables
- enhanced error messages of parser
- enhanced code template infrastructure

Version 0.4.0:
==============
- integrated code library
- optimization and fixes

Version 0.3.0:
==============
- baseline for first hostscope release with wfc code

Version 0.2.0:
==============
- lots of fixes, changes, and enhancements

Version 0.1.0:
==============
- initial version


src/wfc.cc
/*
 *  Copyright (C) 2017-2020, Thomas Maier-Komor
 *
 *  This source file belongs to Wire-Format-Compiler.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "Options.h"
#include "CodeGenerator.h"
#include "XmlGenerator.h"
#include "CodeLibrary.h"
#include "ProtoDriver.h"
#include "io.h"
#include "log.h"
#include "template_lib.h"
#include "version.h"

#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <limits.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

#include <iostream>

using namespace std;

extern void decodeFile(const char *fn);

const char *InstallDir = 0;


void printVersion(ostream &out)
{
	out << "Wire Format Compiler (WFC), Version " VERSION "\nCopyright 2015-2020, Thomas Maier-Komor\n";
}


void printUsage(ostream &out)
{
	out <<	"synopsis: wfc [options] <filename.wfc>\n\n"
		"options:\n"
		"-d <file>   : decode file <file>\n"
		"-f <flag>   : set flag or option <flag> as key=value pair\n"
		"-fwith-<o>  : enable option <o>\n"
		"-fno-<o>    : disable option <o>\n"
		"-f <o=v>    : set option <o> to value <v>\n"
		"-O2         : optimize for speed\n"
		"-Os         : optimize for size\n"
		"-Or         : optimize for review\n"
		"-o <name>   : set output basename to <name>\n"
		"-g          : add comments\n"
		"-l          : generate wfc library only\n"
		"-m <msg>    : generate message <msg>\n"
		"-s          : generate subclasses\n"
		"-t <target> : enable options <target> of .wfc file\n"
		"-p <f>.cct  : postprocess .cct file to .cc for testing\n"
		"-v          : increase verbosity\n"
		"-h          : print help message\n"
		"-V          : print version\n"
		"-I<lib>     : include .cct library or directory\n"
		"\n"
		"flags:\n";
	Options::printHelp(out);
}


#if defined __MINGW32__ || defined __MINGW64__
#define resolve_symlink(b)
#else
void resolve_symlink(char *buf)
{
	struct stat st;
	if (-1 == stat(buf,&st))
		return;
	while ((st.st_mode & S_IFMT) == S_IFLNK) {
		char tmp[PATH_MAX];
		if (-1 == readlink(buf,tmp,sizeof(tmp))) {
			error("unable to resolve link %s: %s",buf,strerror(errno));
			return;
		}
		strcpy(buf,tmp);
		if (-1 == stat(buf,&st))
			return;
	}
}
#endif


static void resolveFromPath(char *fn, const char *arg)
// insecure function!!
#if defined __MINGW32__ || defined __MINGW64__
#define ENVSEP ';'
#else
#define ENVSEP ':'
#endif
{
	struct stat st;
	const char *path = getenv("PATH");
	if (path == 0) {
		strcpy(fn,arg);
		return;
	}
	const char *c = strchr(path,ENVSEP);
	while (c) {
		size_t l = c-path;
		memcpy(fn,path,l);
		if (fn[l-1] != FILESEP) {
			fn[l] = FILESEP;
			++l;
		}
		strcpy(fn+l,arg);
		//diag("looking for %s in %s",arg,fn);
		if (0 == stat(fn,&st)) {
			diag("found %s in %s",arg,fn);
			return;
		}
		path = c+1;
		c = strchr(path,ENVSEP);
	}
	size_t pl = strlen(path);
	memcpy(fn,path,pl);
	if (fn[pl-1] != FILESEP)
		fn[pl++] = FILESEP;
	strcpy(fn+pl,arg);
	//diag("looking for %s in %s",arg,fn);
	if (0 == stat(fn,&st)) {
		diag("found %s in %s",arg,fn);
		return;
	}
#if defined __MINGW32__ || defined __MINGW64__
	//diag("looking for %s in %s",arg,fn);
	if (0 == stat(arg,&st)) {
		diag("found %s in .",arg);
		fn[0] = '.';
		fn[1] = 0;
		return;
	}
#endif
	fn[0] = 0;
}


static void setInstallDir(const char *arg0)
{
	char buf[PATH_MAX];
	size_t n = 0;
#if defined __MINGW32__ || defined __MINGW64__
	if (arg0[1] == ':') {	// we expect an absolute path
#else
	if (arg0[0] == FILESEP) {
#endif
		// absolute path - do nothing
	} else if ((arg0[0] == '.') || strchr(arg0,FILESEP)) {
		// relative path: prefill with cwd
		if (0 == getcwd(buf,sizeof(buf))) {
			buf[0] = '.';
			buf[1] = '/';
			buf[2] = 0;
		} else {
			n = strlen(buf);
			if (buf[n-1] != FILESEP)
				buf[n++] = FILESEP;
		}
	} else {
		// resolve location via path
#if defined __MINGW32__ || defined __MINGW64__
		size_t l = strlen(arg0);
		if (memcmp(arg0+l-4,".exe",4)) {
			char tmp[l+5];
			memcpy(tmp,arg0,l);
			strcpy(tmp+l,".exe");
			resolveFromPath(buf,tmp);
		} else {
			resolveFromPath(buf,arg0);
		}
#else
		resolveFromPath(buf,arg0);
#endif
		n = strlen(buf);
		if (n == 0)
			fatal("unable to resolve %s from PATH",arg0);
	}
	strncpy(buf+n,arg0,sizeof(buf)-n);
	if (char *fs = strrchr(buf,FILESEP)) {
		*fs = 0;
	} else {
		char *wd = getcwd(buf,sizeof(buf));
		assert(wd);
		size_t wl = strlen(wd);
		if (wd[wl-1] != FILESEP) {
			wd[wl] = FILESEP;
			wd[wl+1] = 0;
		}
		InstallDir = strdup(wd);
		diag("InstalDir: %s",InstallDir);
		return;
	}
	char *lfs = strrchr(buf,FILESEP);
	assert(lfs);
	lfs[1] = 0;
	resolve_symlink(buf);
	foldPath(buf);
	InstallDir = strdup(buf);
	diag("installdir: %s",InstallDir);
}


int main(int argc, char *argv[])
{
	vector<string> genMsgs;
	for (int i = 1; i < argc; ++i) {
		if (0 == strcmp("-v",argv[i])) {
#ifdef DEBUG
			if (Diags && !Debug) {
				Debug = true;
				diag("debug messages enabled");
			}
#endif
			if (!Diags) {
				Diags = true;
				diag("diag messages enabled");
			}
		} else if (0 == strcmp("-V",argv[i])) {
			printVersion(cout);
		}
	}
	setInstallDir(argv[0]);
	Options options("commandline",Options::getDefaults());
	bool yydebug = false, hadPostProcess = false, genLib = false;
	const char *target = 0;
	const char *outname = 0;	// output basename to use instead of input file name
	char *fn = 0;
#if defined __MINGW32__ || defined __MINGW64__
#define PCORRECT "+"
#else
#define PCORRECT
	setenv("POSIXLY_CORRECT","1",1);
#endif
	while (optind < argc) {
		int opt = getopt(argc,argv,PCORRECT "d:f:ghI:lm:O:o:p:st:Vvx:y");
		if (opt == -1) {
			if (optind == argc) 
				break;
			if (fn)
				fatal("multiple input files");
			fn = strdup(argv[optind]);
			++optind;
			continue;
		}
		switch (opt) {
		case 'd':
			decodeFile(optarg);
			break;
		case 'f':
			options.addOption(optarg,true);
			break;
		case 'g':
			options.addOption("comments",true);
			break;
		case 'h':
			printUsage(cout);
			exit(EXIT_SUCCESS);
			break;
		case 'I':
			Lib.add(optarg);
			break;
		case 'l':
			genLib = true;
			options.addOption("wfclib","extern",true);
			break;
		case 'm':
			genMsgs.push_back(optarg);
			break;
		case 'O':
			if (0 == strcmp(optarg,"r"))
				options.addOption("Optimize","review",true);
			else if (0 == strcmp(optarg,"2"))
				options.addOption("Optimize","speed",true);
			else if (0 == strcmp(optarg,"s"))
				options.addOption("Optimize","size",true);
			else
				error("invalid argument %s for option -O",optarg);
			break;
		case 'o':
			outname = optarg;
			break;
		case 'p':
			PostProcess(optarg,&options);
			hadPostProcess = true;
			break;
		case 's':
			options.addOption("subclasses",true);
			break;
		case 't':
			target = optarg;
			break;
		case 'V':
			exit(EXIT_SUCCESS);
			break;
		case 'v':
			/*
			 * done at startup
			 * just to prevent "unknown option" messages
			 */
			break;
		case 'x':
			options.addOption(optarg,false);
			break;
		case 'y':
			yydebug = true;
			break;
		default:
			fatal("unknown option -%c",opt);
		}
	}
	string sharedir0 = InstallDir;
	sharedir0 += "share";
	string sharedir1 = sharedir0;
	sharedir1 += FILESEP;
	sharedir1 += "wfc";
	if (0 == access(sharedir1.c_str(),X_OK|R_OK))
		Lib.add(sharedir1.c_str());
	else if (0 == access(sharedir0.c_str(),X_OK|R_OK))
		Lib.add(sharedir0.c_str());
	else
		Lib.addBuf(TemplateLib,"<internal>");
	ProtoDriver drv(yydebug);
	if (fn == 0) {
		if (genLib) {
			CodeGenerator c(0,&options);
			c.writeLib(outname);
			exit(hadError() ? EXIT_FAILURE : EXIT_SUCCESS);
		}
		if (hadPostProcess)
			exit(EXIT_SUCCESS);
		fatal("missing input file");
	}
	msg("parsing %s...",fn);
	drv.parse(fn);
	if (drv.hadError() || hadError())
		exit(EXIT_FAILURE);
	PBFile *f = drv.getFile();
	char *dot = strrchr(fn,'.');
	assert(dot);
	*dot = 0;

	CodeGenerator cg(f,&options);
	if (genLib) {
		cg.writeLib();
		exit(hadError() ? EXIT_FAILURE : EXIT_SUCCESS);
	}
	if (target) {
		cg.setTarget(target);
		if (hadError())
			exit(EXIT_FAILURE);
	}
	cg.init(genMsgs);
	if (hadError())
		exit(EXIT_FAILURE);

	cg.writeFiles(outname ? outname : fn);
	if (NumWarnings)
		msg("Finished with %u warnings.",NumWarnings);
	return 0;
}
